### PR TITLE
chore: update Kex-Algorithms

### PR DESCRIPTION
**Pull request by 34fathombelow**
Signed-off-by: douhunt <douhunt@protonmail.com>

Part 1 of 3 to upgrade base image to Ubuntu:22.04
This **must** be merged before #9551 and cherry-picked into 2.4

Upgraded golang.org/x/crypto libraries to support diffie-hellman-group14-sha256. I also removed two Kex-Algorithms which should no longer be used for security reasons. This may cause some breakage for a very very small group of users. I would also recommend removing diffie-hellman-group14-sha1 in the very near future (v2.5) and give users plenty of warning.
Note on DCO:

If the DCO action in the integration test fails, one or more of your commits are not signed off. Please click on the *Details* link next to the DCO action for instructions on how to resolve this.

Checklist:

* [x] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [ ] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [ ] Optional. My organization is added to USERS.md.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/tree/master/community#contributing-to-argo)
* [ ] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [ ] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).

### CODE REVIEW

**Comment by crenshaw-dev** [comment]
@34fathombelow thanks for this!

Just so we have a paper trail, did you use any particular source that we can cite as the reason for removing the no-longer-secure algorithms?

**Comment by 34fathombelow** [comment]
1. https://tools.ietf.org/id/draft-ietf-curdle-ssh-kex-sha2-09.html#rfc.section.3.6
2. https://github.blog/2017-02-27-crypto-deprecation-notice/

File: util/git/ssh.go [hunk 1]
```diff
@@ -11,13 +11,12 @@ import (
 // Unfortunately, crypto/ssh does not offer public constants or list for
 // this.
 var SupportedSSHKeyExchangeAlgorithms = []string{
```
Comment (crenshaw-dev) [inline util/git/ssh.go:13]: @34fathombelow pointed out that this list should probably be ordered. The crypto/ssh `findCommon` function will pick the first item from this list that is also in the server's list. https://cs.opensource.google/go/x/crypto/+/master:ssh/common.go;l=132?q=findcommon&ss=go%2Fx%2Fcrypto

Suggested order:

```
// Source for strength estimates: https://www.ietf.org/archive/id/draft-ietf-curdle-ssh-kex-sha2-15.html
	"ecdh-sha2-nistp521", // 512 bits
	"ecdh-sha2-nistp384", // 192 bits
	"ecdh-sha2-nistp256", // 128 bits
	"curve25519-sha256@libssh.org", // 128 bits
	// The security strength of the FFC and the hash together will be the minimum of those two values.
	"diffie-hellman-group14-sha256", // 112 bits
	"diffie-hellman-group14-sha1", // 112 bits
	"diffie-hellman-group-exchange-sha256",
```

I couldn't find a source for the strength of `diffie-hellman-group-exchange-sha256`.

**Review by crenshaw-dev — CHANGES_REQUESTED** [review]

File: util/git/ssh.go (hunk 1 above)
Comment (34fathombelow) [reply util/git/ssh.go:13]: I'm not a cryptography expert by any means.  I do believe some of the algorithms come with a performance cost or other complexities that might not scale.   I'll do my best to do some research. 

If anyone has an opinion on this, I'd like to hear it.

Comment (crenshaw-dev) [reply util/git/ssh.go:13]: That's fair. I'd really like to track down a crypto expert to review this change. I'll see what I can do.

Comment (34fathombelow) [reply util/git/ssh.go:13]: Updated the list with all currently [supported](https://cs.opensource.google/go/x/crypto/+/793ad666bf5ec61392092b27061be9618e4e219b:ssh/kex.go) kex-algos with the exception of `diffie-hellman-group1-sha1` & `diffie-hellman-group-exchange-sha1` from the golang.org/x/crypto/ssh package. They are now sorted according to the latest [OpenSSH](https://man.openbsd.org/sshd_config#KexAlgorithms) defaults. The OpenSSH defaults are near identical to Ubuntu:22.04 defaults.

I have ran all unit & e2e test with these changes against 22.04, 21.10, and 20.04

**Review by crenshaw-dev — APPROVED** [review]
LGTM! @jannfis can you take a look?

**Review by jannfis — APPROVED** [review]
LGTM, thanks @34fathombelow !

**Comment by crenshaw-dev** [comment]
Cherry-picked onto 2.4.

**Pull request merged**